File: packages/3d/src/bvh.ts
// A dynamic bounding-volume hierarchy - the physics-broadphase AABB tree
// (Box2D/Bullet lineage): a binary tree over the items' world boxes, NOT a
// subdivision of space. Leaves store FAT boxes (a margin around the tight
// bounds) so an item moving a little refits nothing; an item escaping its
// fat box is removed and re-inserted along the cheapest path (least area
// growth). A ray query walks O(log n) nodes instead of testing every item,
// which is what keeps per-pointer-move picking off the O(meshes) path.
//
// The scene keeps the tree current from its own sync walk: transforms that
// changed are exactly the leaves to update, so maintenance is O(changed)
// per frame - the same delta discipline as rendering, and a static scene
// pays nothing. Storage is flat parallel arrays indexed by node id (no
// per-node objects, no allocation at steady state past tree growth).
//
// Pure module by design: no GUI imports, so the differential check rig
// (checks/pick-check.ts) runs it headless on flux against a linear oracle.

/** Fat-margin fraction of a leaf's largest extent. Bigger = fewer
 * re-inserts while moving, worse query pruning; 5% is the usual trade. */
const MARGIN = 0.05

type Visit<T> = (item: T) => void

export type Bvh<T> = {
  /** Insert an item with its tight world box; returns the leaf handle. */
  insert(item: T, minX: number, minY: number, minZ: number, maxX: number, maxY: number, maxZ: number): number
  /** Update a leaf's tight box. Free while it stays inside the fat box;
   * otherwise the leaf re-inserts. Returns true when it moved. */
  update(leaf: number, minX: number, minY: number, minZ: number, maxX: number, maxY: number, maxZ: number): boolean
  /** Remove a leaf (the handle is dead afterwards). */
  remove(leaf: number): void
  /** Visit every item whose FAT box the ray hits (t >= 0). Broadphase
   * only: the caller narrowphases against its own tight volumes. */
  raycast(ox: number, oy: number, oz: number, dx: number, dy: number, dz: number, visit: Visit<T>): void
}

/**
 * Entry distance of a ray against a box: the smallest t >= 0 with
 * origin + t * direction inside [min, max] (0 when the origin starts
 * inside), or -1 for a miss. The direction need not be normalized - t is
 * in units of its length, which is what keeps a ray transformed into a
 * mesh's local space reporting world distances. Shared by the tree's
 * broadphase and the scene's narrowphase.
 */
export function rayBoxDistance(
  ox: number, oy: number, oz: number,
  dx: number, dy: number, dz: number,
  minX: number, minY: number, minZ: number,
  maxX: number, maxY: number, maxZ: number,
): number {
  let tNear = 0
  let tFar = Infinity
  // Per axis: a zero direction component never crosses the slab, so the
  // origin must already be inside it (the multiply-by-inverse shortcut
  // turns that case into NaN, hence the explicit branch).
  if (dx === 0) {
    if (ox < minX || ox > maxX) return -1
  } else {
    let inv = 1 / dx
    let t1 = (minX - ox) * inv
    let t2 = (maxX - ox) * inv
    if (t1 > t2) { let t = t1; t1 = t2; t2 = t }
    if (t1 > tNear) tNear = t1
    if (t2 < tFar) tFar = t2
  }
  if (dy === 0) {
    if (oy < minY || oy > maxY) return -1
  } else {
    let inv = 1 / dy
    let t1 = (minY - oy) * inv
    let t2 = (maxY - oy) * inv
    if (t1 > t2) { let t = t1; t1 = t2; t2 = t }
    if (t1 > tNear) tNear = t1
    if (t2 < tFar) tFar = t2
  }
  if (dz === 0) {
    if (oz < minZ || oz > maxZ) return -1
  } else {
    let inv = 1 / dz
    let t1 = (minZ - oz) * inv
    let t2 = (maxZ - oz) * inv
    if (t1 > t2) { let t = t1; t1 = t2; t2 = t }
    if (t1 > tNear) tNear = t1
    if (t2 < tFar) tFar = t2
  }
  return tFar >= tNear ? tNear : -1
}

export function createBvh<T>(): Bvh<T> {
  // Node storage, 6 bounds floats per node; child1 === -1 marks a leaf.
  let bounds: number[] = []
  let parent: number[] = []
  let child1: number[] = []
  let child2: number[] = []
  let items: (T | undefined)[] = []
  let free: number[] = []
  let root = -1
  // Traversal stack, reused across queries.
  let stack: number[] = []

  let allocate = (): number => {
    let node = free.pop()
    if (node === undefined) {
      node = parent.length
      bounds.push(0, 0, 0, 0, 0, 0)
      parent.push(-1)
      child1.push(-1)
      child2.push(-1)
      items.push(undefined)
    }
    return node
  }

  // Half the surface area - the SAH cost of a node's box. Degenerate
  // (flat) boxes still cost via their other faces, which is what makes
  // the heuristic sane for planes.
  let area = (n: number): number => {
    let i = n * 6
    let w = bounds[i + 3]! - bounds[i]!
    let h = bounds[i + 4]! - bounds[i + 1]!
    let d = bounds[i + 5]! - bounds[i + 2]!
    return w * (h + d) + h * d
  }

  let unionArea = (n: number, minX: number, minY: number, minZ: number, maxX: number, maxY: number, maxZ: number): number => {
    let i = n * 6
    let w = Math.max(bounds[i + 3]!, maxX) - Math.min(bounds[i]!, minX)
    let h = Math.max(bounds[i + 4]!, maxY) - Math.min(bounds[i + 1]!, minY)
    let d = Math.max(bounds[i + 5]!, maxZ) - Math.min(bounds[i + 2]!, minZ)
    return w * (h + d) + h * d
  }

  // Recompute an internal node's box as the union of its children.
  let refit = (n: number): void => {
    let a = child1[n]! * 6
    let b = child2[n]! * 6
    let i = n * 6
    bounds[i] = Math.min(bounds[a]!, bounds[b]!)
    bounds[i + 1] = Math.min(bounds[a + 1]!, bounds[b + 1]!)
    bounds[i + 2] = Math.min(bounds[a + 2]!, bounds[b + 2]!)
    bounds[i + 3] = Math.max(bounds[a + 3]!, bounds[b + 3]!)
    bounds[i + 4] = Math.max(bounds[a + 4]!, bounds[b + 4]!)
    bounds[i + 5] = Math.max(bounds[a + 5]!, bounds[b + 5]!)
  }

  let insertLeaf = (leaf: number): void => {
    if (root === -1) {
      root = leaf
      parent[leaf] = -1
      return
    }
    let i = leaf * 6
    let minX = bounds[i]!, minY = bounds[i + 1]!, minZ = bounds[i + 2]!
    let maxX = bounds[i + 3]!, maxY = bounds[i + 4]!, maxZ = bounds[i + 5]!
    // Descend toward the sibling that grows the least (the classic
    // incremental surface-area heuristic).
    let sibling = root
    while (child1[sibling] !== -1) {
      let a = child1[sibling]!
      let b = child2[sibling]!
      let costA = unionArea(a, minX, minY, minZ, maxX, maxY, maxZ) - area(a)
      let costB = unionArea(b, minX, minY, minZ, maxX, maxY, maxZ) - area(b)
      sibling = costA < costB ? a : b
    }
    let oldParent = parent[sibling]!
    let newParent = allocate()
    parent[newParent] = oldParent
    child1[newParent] = sibling
    child2[newParent] = leaf
    items[newParent] = undefined
    parent[sibling] = newParent
    parent[leaf] = newParent
    if (oldParent === -1) {
      root = newParent
    } else if (child1[oldParent] === sibling) {
      child1[oldParent] = newParent
    } else {
      child2[oldParent] = newParent
    }
    for (let n = newParent; n !== -1; n = parent[n]!) refit(n)
  }

  let removeLeaf = (leaf: number): void => {
    if (leaf === root) {
      root = -1
      return
    }
    let p = parent[leaf]!
    let sibling = child1[p] === leaf ? child2[p]! : child1[p]!
    let grand = parent[p]!
    parent[sibling] = grand
    if (grand === -1) {
      root = sibling
    } else {
      if (child1[grand] === p) child1[grand] = sibling
      else child2[grand] = sibling
      for (let n = grand; n !== -1; n = parent[n]!) refit(n)
    }
    child1[p] = -1
    items[p] = undefined
    free.push(p)
  }

  let setFat = (leaf: number, minX: number, minY: number, minZ: number, maxX: number, maxY: number, maxZ: number): void => {
    let m = MARGIN * Math.max(maxX - minX, maxY - minY, maxZ - minZ)
    let i = leaf * 6
    bounds[i] = minX - m
    bounds[i + 1] = minY - m
    bounds[i + 2] = minZ - m
    bounds[i + 3] = maxX + m
    bounds[i + 4] = maxY + m
    bounds[i + 5] = maxZ + m
  }

  return {
    insert(item, minX, minY, minZ, maxX, maxY, maxZ) {
      let leaf = allocate()
      items[leaf] = item
      setFat(leaf, minX, minY, minZ, maxX, maxY, maxZ)
      insertLeaf(leaf)
      return leaf
    },
    update(leaf, minX, minY, minZ, maxX, maxY, maxZ) {
      let i = leaf * 6
      if (
        bounds[i]! <= minX && bounds[i + 1]! <= minY && bounds[i + 2]! <= minZ &&
        bounds[i + 3]! >= maxX && bounds[i + 4]! >= maxY && bounds[i + 5]! >= maxZ
      ) {
        return false
      }
      removeLeaf(leaf)
      setFat(leaf, minX, minY, minZ, maxX, maxY, maxZ)
      insertLeaf(leaf)
      return true
    },
    remove(leaf) {
      removeLeaf(leaf)
      items[leaf] = undefined
      free.push(leaf)
    },
    raycast(ox, oy, oz, dx, dy, dz, visit) {
      if (root === -1) return
      stack.length = 0
      stack.push(root)
      while (stack.length > 0) {
        let n = stack.pop()!
        let i = n * 6
        let t = rayBoxDistance(ox, oy, oz, dx, dy, dz, bounds[i]!, bounds[i + 1]!, bounds[i + 2]!, bounds[i + 3]!, bounds[i + 4]!, bounds[i + 5]!)
        if (t < 0) continue
        if (child1[n] === -1) {
          visit(items[n] as T)
        } else {
          stack.push(child1[n]!, child2[n]!)
        }
      }
    },
  }
}

File: alloy/src/spatial/bvh.rs
// A dynamic bounding-volume hierarchy - the physics-broadphase AABB tree
// (Box2D/Bullet lineage): a binary tree over the items' world boxes, NOT a
// subdivision of space. Leaves store FAT boxes (a margin around the tight
// bounds) so an item moving a little refits nothing; an item escaping its
// fat box is removed and re-inserted along the cheapest path (least area
// growth). A ray query walks O(log n) nodes instead of testing every item.
// Flat parallel storage indexed by tree node; the item is the caller's
// u32 (a spatial node index).

/// Fat-margin fraction of a leaf's largest extent. Bigger = fewer
/// re-inserts while moving, worse query pruning; 5% is the usual trade.
const MARGIN: f32 = 0.05;

pub type Box3 = [f32; 6];

pub struct Bvh {
  bounds: Vec<Box3>,
  parent: Vec<i32>,
  child1: Vec<i32>,
  child2: Vec<i32>,
  items: Vec<u32>,
  free: Vec<i32>,
  root: i32,
  stack: Vec<i32>,
}

/// Entry distance of a ray against a box: the smallest t >= 0 with
/// origin + t * direction inside the box (0 when the origin starts
/// inside), or None for a miss. The direction need not be normalized - t
/// is in units of its length, which is what keeps a ray transformed into
/// a mesh's local space reporting world distances.
pub fn ray_box_distance(o: [f32; 3], d: [f32; 3], b: &Box3) -> Option<f32> {
  let mut t_near = 0.0f32;
  let mut t_far = f32::INFINITY;
  for axis in 0..3 {
    // A zero direction component never crosses the slab, so the origin
    // must already be inside it (the inverse shortcut would yield NaN).
    if d[axis] == 0.0 {
      if o[axis] < b[axis] || o[axis] > b[axis + 3] {
        return None;
      }
    } else {
      let inv = 1.0 / d[axis];
      let mut t1 = (b[axis] - o[axis]) * inv;
      let mut t2 = (b[axis + 3] - o[axis]) * inv;
      if t1 > t2 {
        std::mem::swap(&mut t1, &mut t2);
      }
      t_near = t_near.max(t1);
      t_far = t_far.min(t2);
    }
  }
  if t_far >= t_near {
    Some(t_near)
  } else {
    None
  }
}

fn half_area(b: &Box3) -> f32 {
  let w = b[3] - b[0];
  let h = b[4] - b[1];
  let d = b[5] - b[2];
  w * (h + d) + h * d
}

fn union(a: &Box3, b: &Box3) -> Box3 {
  [a[0].min(b[0]), a[1].min(b[1]), a[2].min(b[2]), a[3].max(b[3]), a[4].max(b[4]), a[5].max(b[5])]
}

fn fatten(b: &Box3) -> Box3 {
  let m = MARGIN * (b[3] - b[0]).max(b[4] - b[1]).max(b[5] - b[2]);
  [b[0] - m, b[1] - m, b[2] - m, b[3] + m, b[4] + m, b[5] + m]
}

impl Default for Bvh {
  fn default() -> Self {
    Bvh {
      bounds: Vec::new(),
      parent: Vec::new(),
      child1: Vec::new(),
      child2: Vec::new(),
      items: Vec::new(),
      free: Vec::new(),
      root: -1,
      stack: Vec::new(),
    }
  }
}

impl Bvh {
  fn allocate(&mut self) -> i32 {
    if let Some(n) = self.free.pop() {
      return n;
    }
    self.bounds.push([0.0; 6]);
    self.parent.push(-1);
    self.child1.push(-1);
    self.child2.push(-1);
    self.items.push(0);
    (self.bounds.len() - 1) as i32
  }

  fn refit(&mut self, n: i32) {
    let a = self.bounds[self.child1[n as usize] as usize];
    let b = self.bounds[self.child2[n as usize] as usize];
    self.bounds[n as usize] = union(&a, &b);
  }

  fn insert_leaf(&mut self, leaf: i32) {
    if self.root == -1 {
      self.root = leaf;
      self.parent[leaf as usize] = -1;
      return;
    }
    let lb = self.bounds[leaf as usize];
    // Descend toward the sibling that grows the least (the classic
    // incremental surface-area heuristic).
    let mut sibling = self.root;
    while self.child1[sibling as usize] != -1 {
      let a = self.child1[sibling as usize];
      let b = self.child2[sibling as usize];
      let ba = &self.bounds[a as usize];
      let bb = &self.bounds[b as usize];
      let cost_a = half_area(&union(ba, &lb)) - half_area(ba);
      let cost_b = half_area(&union(bb, &lb)) - half_area(bb);
      sibling = if cost_a < cost_b { a } else { b };
    }
    let old_parent = self.parent[sibling as usize];
    let new_parent = self.allocate();
    self.parent[new_parent as usize] = old_parent;
    self.child1[new_parent as usize] = sibling;
    self.child2[new_parent as usize] = leaf;
    self.parent[sibling as usize] = new_parent;
    self.parent[leaf as usize] = new_parent;
    if old_parent == -1 {
      self.root = new_parent;
    } else if self.child1[old_parent as usize] == sibling {
      self.child1[old_parent as usize] = new_parent;
    } else {
      self.child2[old_parent as usize] = new_parent;
    }
    let mut n = new_parent;
    while n != -1 {
      self.refit(n);
      self.rotate(n);
      n = self.parent[n as usize];
    }
  }

  // One SAH rotation at n (Box2D lineage): try swapping each child with a
  // grandchild from the other side and keep the swap that shrinks the
  // rotated child's box the most. Applied along every refit walk, this
  // keeps the tree shallow under adversarial insertion orders (a grid
  // inserted row by row degenerates a rotation-free SAH tree into deep
  // chains) at a constant cost per walked node.
  fn rotate(&mut self, n: i32) {
    let c1 = self.child1[n as usize];
    let c2 = self.child2[n as usize];
    if c1 == -1 {
      return;
    }
    // (gain, rotated child, its kept grandchild slot, the swapped-in node)
    let mut best: Option<(f32, i32, bool, i32)> = None;
    let mut consider = |rotated: i32, swapped_in: i32, bvh: &Bvh| {
      if bvh.child1[rotated as usize] == -1 {
        return;
      }
      let g1 = bvh.child1[rotated as usize];
      let g2 = bvh.child2[rotated as usize];
      let current = half_area(&bvh.bounds[rotated as usize]);
      let sb = &bvh.bounds[swapped_in as usize];
      // Swap `swapped_in` with g2 (keeping g1), then with g1 (keeping g2).
      for (keep_first, kept) in [(true, g1), (false, g2)] {
        let gain = current - half_area(&union(&bvh.bounds[kept as usize], sb));
        if gain > 0.0 && best.is_none_or(|(g, ..)| gain > g) {
          best = Some((gain, rotated, keep_first, swapped_in));
        }
      }
    };
    consider(c1, c2, self);
    consider(c2, c1, self);
    let Some((_, rotated, keep_first, swapped_in)) = best else {
      return;
    };
    let dropped = if keep_first { self.child2[rotated as usize] } else { self.child1[rotated as usize] };
    // `swapped_in` (the other child of n) takes the dropped grandchild's
    // slot; the dropped grandchild becomes n's direct child.
    if keep_first {
      self.child2[rotated as usize] = swapped_in;
    } else {
      self.child1[rotated as usize] = swapped_in;
    }
    self.parent[swapped_in as usize] = rotated;
    if self.child1[n as usize] == swapped_in {
      self.child1[n as usize] = dropped;
    } else {
      self.child2[n as usize] = dropped;
    }
    self.parent[dropped as usize] = n;
    self.refit(rotated);
    self.refit(n);
  }

  /// Longest root-to-leaf path (tests: tree-quality assertions).
  #[cfg(test)]
  pub(crate) fn depth(&self) -> usize {
    fn walk(bvh: &Bvh, n: i32) -> usize {
      if n == -1 || bvh.child1[n as usize] == -1 {
        return 0;
      }
      1 + walk(bvh, bvh.child1[n as usize]).max(walk(bvh, bvh.child2[n as usize]))
    }
    walk(self, self.root)
  }

  fn remove_leaf(&mut self, leaf: i32) {
    if leaf == self.root {
      self.root = -1;
      return;
    }
    let p = self.parent[leaf as usize];
    let sibling = if self.child1[p as usize] == leaf { self.child2[p as usize] } else { self.child1[p as usize] };
    let grand = self.parent[p as usize];
    self.parent[sibling as usize] = grand;
    if grand == -1 {
      self.root = sibling;
    } else {
      if self.child1[grand as usize] == p {
        self.child1[grand as usize] = sibling;
      } else {
        self.child2[grand as usize] = sibling;
      }
      let mut n = grand;
      while n != -1 {
        self.refit(n);
        self.rotate(n);
        n = self.parent[n as usize];
      }
    }
    self.child1[p as usize] = -1;
    self.free.push(p);
  }

  /// Insert an item with its tight world box; returns the leaf handle.
  pub fn insert(&mut self, item: u32, tight: &Box3) -> u32 {
    let leaf = self.allocate();
    self.items[leaf as usize] = item;
    self.child1[leaf as usize] = -1;
    self.bounds[leaf as usize] = fatten(tight);
    self.insert_leaf(leaf);
    leaf as u32
  }

  /// Update a leaf's tight box: free while it stays inside the fat box,
  /// otherwise the leaf re-inserts.
  pub fn update(&mut self, leaf: u32, tight: &Box3) {
    let fat = &self.bounds[leaf as usize];
    if fat[0] <= tight[0]
      && fat[1] <= tight[1]
      && fat[2] <= tight[2]
      && fat[3] >= tight[3]
      && fat[4] >= tight[4]
      && fat[5] >= tight[5]
    {
      return;
    }
    let leaf = leaf as i32;
    self.remove_leaf(leaf);
    self.bounds[leaf as usize] = fatten(tight);
    self.insert_leaf(leaf);
  }

  /// Remove a leaf (the handle is dead afterwards).
  pub fn remove(&mut self, leaf: u32) {
    self.remove_leaf(leaf as i32);
    self.free.push(leaf as i32);
  }

  /// Visit every item whose FAT box the ray hits. Broadphase only: the
  /// caller narrowphases against its own tight volumes.
  pub fn raycast(&mut self, o: [f32; 3], d: [f32; 3], visit: &mut dyn FnMut(u32)) {
    if self.root == -1 {
      return;
    }
    let mut stack = std::mem::take(&mut self.stack);
    stack.clear();
    stack.push(self.root);
    while let Some(n) = stack.pop() {
      if ray_box_distance(o, d, &self.bounds[n as usize]).is_none() {
        continue;
      }
      if self.child1[n as usize] == -1 {
        visit(self.items[n as usize]);
      } else {
        stack.push(self.child1[n as usize]);
        stack.push(self.child2[n as usize]);
      }
    }
    self.stack = stack;
  }
}
